### PR TITLE
kubeadm: Add a `v1alpha3` API

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
Adds a (now duplicated) v1alpha3 API. Equal to `v1alpha2` now, but will be changed later.
Hopefully we'll graduate this to `v1beta1` during the v1.12 cycle.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
ref: https://github.com/kubernetes/kubeadm/issues/911
Depends on:
 - [x] https://github.com/kubernetes/kubernetes/pull/65776
 - [x] https://github.com/kubernetes/kubernetes/pull/65628

**Special notes for your reviewer**:

These commits should be reviewed:
 - `Add a duplicated v1alpha3 API`: A pure directory copy of the `v1alpha2` folder, only changed the package name in the Go files
 - `Register the v1alpha3 API in the scheme, and update the roundtrip API tests`, started registering the new API and using it internally
 - `Automated bump from v1alpha2 references to v1alpha3`: Replaced all v1alpha2 references with v1alpha3

**Release note**:

```release-note
kubeadm: Add a `v1alpha3` API.
```
@kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

**Review by fabriziopandini — COMMENTED** [review]
Thanks for this PR
/approve
@timothysc I let you give the final lgtm

**Comment by timothysc** [comment]
let me know when all the rebasing is done.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/bootstraptokenstring.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package v1alpha2 holds the external kubeadm API types of version v1alpha2
+// Note: This file should be kept in sync with the similar one for the internal API
+// TODO: The BootstrapTokenString object should move out to either k8s.io/client-go or k8s.io/api in the future
```
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha3/bootstraptokenstring.go]: Now is the time to make this change, it really doesn't belong here.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/defaults_unix.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+// DefaultCACertPath defines default location of CA certificate on Linux
+const DefaultCACertPath = "/etc/kubernetes/pki/ca.crt"
```
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha3/defaults_unix.go]: Why is this outside of the rest of the other defaults.

**Review by timothysc — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/bootstraptokenstring.go (hunk 1 above)
Comment (luxas) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha3/bootstraptokenstring.go]: This is out of scope for this PR, it isn't worse than what we had with v1alpha1 and v1alpha2 really.
Yes, we're gonna fix this very soon, but it doesn't block all of the other PRs I have lined up for the moment

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/defaults_unix.go (hunk 2 above)
Comment (luxas) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha3/defaults_unix.go]: It's branched per OS, there's an other constant for windows.

Comment (luxas) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha3/defaults_unix.go]: Again, this is 100% equal to what's in v1alpha2

**Comment by fabriziopandini** [comment]
This PR sets the ground for the new version of kubeadm API, but doesn't make any change to API types (to be addressed in following PRs along with the related KEP)
/lgtm
/approve

**Comment by luxas** [comment]
Self-approving the `.golint_failures` change. We have that for our `v1alpha2` API as well, (due to the apimachinery defaulting code for example)

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/bootstraptokenstring.go (hunk 1 above)
Comment (timothysc) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha3/bootstraptokenstring.go]: Please file an issue in the kubeadm repo as p1 as it needs to get fixed this cycle. 

/cc @fabriziopandini

Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha3/bootstraptokenstring.go]: I added an action item to https://github.com/kubernetes/kubeadm/issues/963 Tracking issue for "Config to v1beta1"

**Pull request merged**